File: boj/java/2839-a.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Main {


    public void bojSolution() throws Exception {
        // 입력
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        br.close();

        int start = n / 5;
        List<Integer> cntList = new ArrayList<>(start + 1);
        for (int i = start; i >= 0; i--) {
            int goal = n - 5 * i;
            if (goal % 3 == 0) {
                cntList.add(i + goal / 3);
            }
        }

        // 출력
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        if (cntList.isEmpty()) {
            bw.write(String.valueOf(-1));
        } else {
            bw.write(String.valueOf(Collections.min(cntList)));
        }
        bw.flush();
        bw.close();
    }

    public static void main(String[] args) throws Exception {
        new Main().bojSolution();
    }
}

File: boj/cpp/2839-a.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
int main()
{
    cin >> n;

    // 5로 나눠서 시작점을 계산하자.
    int start = n / 5;

    // 몇 봉지인지 계산결과를 저장
    vector<int> cnt_list;
    for (int i = start; i >= 0; i--)
    {
        // 목표무게 : n
        int goal = n - i * 5;
        // 5 * i 를 제외하고 3으로 나눠지는지 여부 확인하면 정확하게 N그램 배달하는지 알 수 있다.
        if (goal % 3 == 0)
        {
            cnt_list.push_back(i + goal / 3);
        }
    }
    // 못 찾는 경우
    if (cnt_list.empty())
    {
        cout << -1 << "\n";
        return 0;
    }
    cout << *min_element(cnt_list.begin(), cnt_list.end());
    return 0;
}